### PR TITLE
MusicGen: Add Stereo Model

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

The original MusicGen model generates mono (1-channel) outputs. It does this by predicting a set of 4 codebooks at each generation step:
```
[codebook_1, codebook_2, codebook_3, codebook_4]
```
After generating, the sequence of predicted codebooks is passed through the EnCodec model to get the final waveform.

This PR adds the MusicGen **stereo** model. It works by predicting **two** sets of codebooks at each step. One set of codebooks corresponds to the left channel, the other set corresponds to the right channel. The sets of codebooks are interleaved as follows:
```
[left_codebook_1, right_codebook_1, left_codebook_2, right_codebook_2, ..., left_codebook_4, right_codebook_4]
```

After generating, the sequence of generated codebooks are partitioned into their left/right parts, and then each sequence passed through EnCodec to get the left/right waveform respectively.

### CODE REVIEW

File: src/transformers/models/musicgen/configuration_musicgen.py [hunk 1]
```diff
@@ -75,6 +75,9 @@ class MusicgenDecoderConfig(PretrainedConfig):
             The number of parallel codebooks forwarded to the model.
         tie_word_embeddings(`bool`, *optional*, defaults to `False`):
             Whether input and output word embeddings should be tied.
+        audio_channels (`int`, *optional*, defaults to 1
```
Comment (sanchit-gandhi) [inline src/transformers/models/musicgen/configuration_musicgen.py:78]: Following the EnCodec naming here: https://github.com/huggingface/transformers/blob/d7cb5e138ec1ccc848a554574b1a89f0dfaf0e90/src/transformers/models/encodec/configuration_encodec.py#L50-L51

Note that the EnCodec model used is still 1-channel (mono) - it's just the MusicGen model that works in a 2-channel fashion.

**Review by ArthurZucker — COMMENTED** [review]
Thanks looks very clean 😉

**Review by ArthurZucker — APPROVED** [review]

**Comment by ArthurZucker** [comment]
Is the format we gave (interleaving) forced by BC? Otherwise storing in two list or tuples would be better IMO (kind of like how audio are stored no?) 
my only "complain" here

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 2]
```diff
@@ -1138,21 +1138,33 @@ def build_delay_pattern_mask(self, input_ids: torch.LongTensor, pad_token_id: in
...
         delay_pattern = torch.triu(
-            torch.ones((num_codebooks, max_length), dtype=torch.bool), diagonal=max_length - num_codebooks + 1
+            torch.ones((channel_codebooks, max_length), dtype=torch.bool), diagonal=max_length - channel_codebooks + 1
         )
         # then fill the lower triangular part (the BOS padding)
-        delay_pattern = delay_pattern + torch.tril(torch.ones((num_codebooks, max_length), dtype=torch.bool))
+        delay_pattern = delay_pattern + torch.tril(torch.ones((channel_codebooks, max_length), dtype=torch.bool))
+
+        if self.config.audio_channels == 2:
+            # for left/right channel we need to duplicate every row of the pattern pask in an interleaved fashion
```
Comment (sanchit-gandhi) [inline src/transformers/models/musicgen/modeling_musicgen.py:1165]:
```suggestion
            # for left/right channel we need to duplicate every row of the pattern mask in an interleaved fashion
```

[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/musicgen/modeling_musicgen.py]: Resolved in https://github.com/huggingface/transformers/pull/27084/commits/02ae8f66b59a7b469e3ee4fde08f552004d7a968

File: src/transformers/models/musicgen/modeling_musicgen.py [hunk 3]
```diff
@@ -2499,16 +2546,29 @@ def generate(
...
+        else:
+            output_values_left = self.audio_encoder.decode(
+                output_ids[:, :, ::2, :],
+                audio_scales=audio_scales,
+            ).audio_values
+
+            output_values_right = self.audio_encoder.decode(
+                output_ids[:, :, 1::2, :],
+                audio_scales=audio_scales,
+            ).audio_values
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/musicgen/modeling_musicgen.py]: Resolved in https://github.com/huggingface/transformers/pull/27084/commits/fa300604670af8d732c4c72856abd15f27261e16

**Comment by sanchit-gandhi** [comment]
The original model is designed to predict them in an interleaved way:
```
[left_codebook_1, right_codebook_1, left_codebook_2, right_codebook_2, ..., left_codebook_4, right_codebook_4]
```
We could change this to predict left first, then right:
```
[left_codebook_1, left_codebook_2, ..., left_codebook_4, right_codebook_1, right_codebook_2, ..., right_codebook_4]
```
Which would require re-shaping the LM head weights, and duplicating the pattern mask along the row dimension. Overall I think the complexity would be similar to the interleaved way we have now.

But predicting two sets of codebooks as two separate tuples would break compatibility with the existing mono musicgen, or otherwise complicate the code since we'll have different inputs / sampling logic depending on whether we're mono / stereo.

**Comment by ArthurZucker** [comment]
Awesome thanks for explaining
